src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Embedded.cs

```diff
@@ -0,0 +1,435 @@
+        {
+            var reference = CreateCompilationWithMscorlib(@"
+namespace Microsoft.CodeAnalysis
+{
+    internal class EmbeddedAttribute : System.Attribute { }
+}
+namespace TestReference
+{
+    [Microsoft.CodeAnalysis.Embedded]
+    public class TestType1 { }
+
+    [Microsoft.CodeAnalysis.EmbeddedAttribute]
+    public class TestType2 { }
+
+    public class TestType3 { }
+}");
+
+            var code = @"
+class Program
+{
+    public static void Main()
+    {
+        var obj1 = new TestReference.TestType1();
+        var obj2 = new TestReference.TestType2();
+        var obj3 = new TestReference.TestType3(); // This should be fine
+    }
+}";
+
+            CreateCompilationWithMscorlib(code, references: new[] { reference.ToMetadataReference() }).VerifyDiagnostics(
+                // (6,38): error CS0234: The type or namespace name 'TestType1' does not exist in the namespace 'TestReference' (are you missing an assembly reference?)
+                //         var obj1 = new TestReference.TestType1();
+                Diagnostic(ErrorCode.ERR_DottedTypeNameNotFoundInNS, "TestType1").WithArguments("TestType1", "TestReference").WithLocation(6, 38),
+                // (7,38): error CS0234: The type or namespace name 'TestType2' does not exist in the namespace 'TestReference' (are you missing an assembly reference?)
+                //         var obj2 = new TestReference.TestType2();
+                Diagnostic(ErrorCode.ERR_DottedTypeNameNotFoundInNS, "TestType2").WithArguments("TestType2", "TestReference").WithLocation(7, 38));
+        }
+
+        [Fact]
+        public void EmbeddedAttributeInSourceIsAllowedIfCompilerDoesNotNeedToGenerateOne()
+        {
+            var code = @"
+namespace Microsoft.CodeAnalysis
+{
+    public class EmbeddedAttribute : System.Attribute { }
+}
+namespace OtherNamespace
+{
+    [Microsoft.CodeAnalysis.EmbeddedAttribute]
+    public class TestReference
+    {
+        public static int GetValue() => 3;
+    }
+}
+class Test
+{
+    public static void Main()
+    {
+        // This should be fine, as the compiler doesn't need to use an embedded attribute for this compilation
+        System.Console.Write(OtherNamespace.TestReference.GetValue());
+    }
+}";
+
+            CompileAndVerify(code, verify: false, expectedOutput: "3");
+        }
+
+        [Fact]
+        public void EmbeddedAttributeInSourceShouldTriggerAnErrorIfCompilerNeedsToGenerateOne()
+        {
+            var code = @"
+namespace Microsoft.CodeAnalysis
+{
+    public class EmbeddedAttribute : System.Attribute { }
+}
+class Test
+{
+    public void M(ref readonly int p)
+    {
+        // This should trigger generating another EmbeddedAttribute
+    }
+}";
+
+            CreateCompilationWithMscorlib(code, assemblyName: "testModule").VerifyEmitDiagnostics(
+                // (4,18): error CS8413: The type name 'EmbeddedAttribute' is reserved to be used by the compiler.
```

Comment:
Consider reporting the fully qualified type name so it's clear we're not reserving the type name regardless of namespace.